<a href="https://colab.research.google.com/github/sforceas/machine-learning/blob/master/BreastCancer_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [98]:
import pandas as pd
import numpy as np
import scipy.stats as ss
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
import sklearn.metrics as metrics

# Wisconsin Breast Cancer Dataset
Description: Features are computed from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image.


Attribute information

1) ID number
2) Diagnosis (M = malignant, B = benign)
3-32)

Ten real-valued features are computed for each cell nucleus:

	a) radius (mean of distances from center to points on the perimeter)
	b) texture (standard deviation of gray-scale values)
	c) perimeter
	d) area
	e) smoothness (local variation in radius lengths)
	f) compactness (perimeter^2 / area - 1.0)
	g) concavity (severity of concave portions of the contour)
	h) concave points (number of concave portions of the contour)
	i) symmetry 
	j) fractal dimension ("coastline approximation" - 1)

Several of the papers listed above contain detailed descriptions of
how these features are computed. 

The mean, standard error, and "worst" or largest (mean of the three
largest values) of these features were computed for each image,
resulting in 30 features.  For instance, field 3 is Mean Radius, field
13 is Radius SE, field 23 is Worst Radius.


https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic)

In [99]:
#cabezeras de la base de datos en el repositorio de la UCI. SE = Standard Error, W = Worst Values
headers = ['id','type','radius','texture','perimeter','area','smoothness','compactness','concavity','concave-points','symmetry','fractal dimension','SE_radius','SE_texture','SE_perimeter','SE_area','SE_smoothness','SE_compactness','SE_concavity','SE_concave-points','SE_symmetry','SE_fractal dimension','W_radius','W_texture','W_perimeter','W_area','W_smoothness','W_compactness','W_concavity','W_concave-points','W_symmetry','W_fractal dimension']

In [100]:
#URL de la base de datos en el repositorio de la UCI
db_url="https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data"


In [101]:
df= pd.read_csv(db_url,names=headers)
df

id type  radius  ...  W_concave-points  W_symmetry  W_fractal dimension
0      842302    M   17.99  ...            0.2654      0.4601              0.11890
1      842517    M   20.57  ...            0.1860      0.2750              0.08902
2    84300903    M   19.69  ...            0.2430      0.3613              0.08758
3    84348301    M   11.42  ...            0.2575      0.6638              0.17300
4    84358402    M   20.29  ...            0.1625      0.2364              0.07678
..        ...  ...     ...  ...               ...         ...                  ...
564    926424    M   21.56  ...            0.2216      0.2060              0.07115
565    926682    M   20.13  ...            0.1628      0.2572              0.06637
566    926954    M   16.60  ...            0.1418      0.2218              0.07820
567    927241    M   20.60  ...            0.2650      0.4087              0.12400
568     92751    B    7.76  ...            0.0000      0.2871              0.07039

[569 rows x 32 columns]

# Model 1. Mean values as features

## Spliting Train and Test datasets

In [102]:
feature_cols = ['radius','texture','perimeter','area','smoothness','compactness','concavity','concave-points','symmetry','fractal dimension']

In [103]:
x = df[feature_cols]
y = df.type

In [104]:
X_train,X_test,y_train,y_test = train_test_split(x,y, test_size=0.3, random_state=0)

In [105]:
y_train

478    B
303    B
155    B
186    M
101    B
      ..
277    M
9      M
359    B
192    B
559    B
Name: type, Length: 398, dtype: object

## Model Building

In [106]:
logreg = LogisticRegression()
logreg.fit(X_train,y_train)
y_pred = logreg.predict(X_test)
y_pred

array(['B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'M', 'B',
       'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B',
       'M', 'M', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'M',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'M',
       'B', 'M', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B',
       'M', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M',
       'M', 'B', 'M', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B',
       'B', 'B'], dtype=object)

## Confusion matrix and accuracy


In [107]:
cnf_matrix = pd.crosstab(y_test ,y_pred ,rownames=['Real'], colnames=['Prediction'])
cnf_matrix

Prediction    B   M
Real               
B           102   6
M             8  55

In [108]:
accurracy_M = metrics.accuracy_score(y_test,y_pred)
print(f'Accuracy = {round(accurracy*100,1)}%')

Accuracy = 91.8%


# Model 2. Worst values as features


## Spliting Train and Test datasets

In [109]:
feature_cols = ['W_radius','W_texture','W_perimeter','W_area','W_smoothness','W_compactness','W_concavity','W_concave-points','W_symmetry','W_fractal dimension']

In [110]:
x = df[feature_cols]
y = df.type

In [111]:
X_train,X_test,y_train,y_test = train_test_split(x,y, test_size=0.3, random_state=0)

In [112]:
y_train

478    B
303    B
155    B
186    M
101    B
      ..
277    M
9      M
359    B
192    B
559    B
Name: type, Length: 398, dtype: object

## Model Building

In [127]:
logreg = LogisticRegression(solver='lbfgs', max_iter=2000)
logreg.fit(X_train,y_train)
y_pred = logreg.predict(X_test)
y_pred

array(['M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'M', 'B',
       'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B',
       'M', 'M', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'M', 'M', 'M',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'M',
       'B', 'M', 'M', 'M', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B',
       'M', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'B', 'M', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M', 'M', 'M', 'B',
       'B', 'B'], dtype=object)

## Confusion matrix and accuracy


In [128]:
cnf_matrix = pd.crosstab(y_test ,y_pred ,rownames=['Real'], colnames=['Prediction'])
cnf_matrix

Prediction    B   M
Real               
B           102   6
M             1  62

In [130]:
accurracy = metrics.accuracy_score(y_test,y_pred)
print(f'Accuracy = {round(accurracy*100,1)}%')

Accuracy = 95.9%


# Model 3. Mean & Worst values as features


## Spliting Train and Test datasets

In [116]:
feature_cols = ['radius','texture','perimeter','area','smoothness','compactness','concavity','concave-points','symmetry','fractal dimension','W_radius','W_texture','W_perimeter','W_area','W_smoothness','W_compactness','W_concavity','W_concave-points','W_symmetry','W_fractal dimension']

In [117]:
x = df[feature_cols]
y = df.type

In [118]:
X_train,X_test,y_train,y_test = train_test_split(x,y, test_size=0.3, random_state=0)

In [119]:
y_train

478    B
303    B
155    B
186    M
101    B
      ..
277    M
9      M
359    B
192    B
559    B
Name: type, Length: 398, dtype: object

## Model Building

In [131]:
logreg = LogisticRegression(solver='lbfgs', max_iter=2000)
logreg.fit(X_train,y_train)
y_pred = logreg.predict(X_test)
y_pred

array(['M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'M', 'B',
       'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B',
       'M', 'M', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'M', 'M', 'M',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'M',
       'B', 'M', 'M', 'M', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B',
       'M', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'M', 'B', 'M', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'B', 'M', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M', 'M', 'M', 'B',
       'B', 'B'], dtype=object)

## Confusion matrix and accuracy


In [132]:
cnf_matrix = pd.crosstab(y_test ,y_pred ,rownames=['Real'], colnames=['Prediction'])
cnf_matrix

Prediction    B   M
Real               
B           102   6
M             1  62

In [134]:
accurracy = metrics.accuracy_score(y_test,y_pred)
print(f'Accuracy = {round(accurracy*100,1)}%')

Accuracy = 95.9%


# Conclussion
Better prediction accuracy with less features by using the model 2: 95.9% accuracy score and only 1 Fake Negative.